### PR TITLE
[fix/passcode-device-enforced] MDM Parameter - Enforce Passcode Setup

### PR DESCRIPTION
**Pull request by hosy**
## Description
This PR implements a new MDM parameter `passcode.enforced-by-device`. If this value is `true` a passcode is required, if NO device passcode is set.

This PR further fixes a bug where the alert could be dismissed without action on the iPad, even though the parameter was enforced.

## Related Issue
https://github.com/owncloud/enterprise/issues/5526

## How Has This Been Tested?

In `Branding.plist` set the following value:

```
	<key>passcode.enforced-by-device</key>
	<true/>
``` 

Test with device passcode enabled/disabled.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] Added an issue with details about all relevant changes in the [**iOS documentation repository**](https://github.com/owncloud/docs-client-ios-app/issues).
- [ ] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added changelog files for the fixed issues in folder changelog/unreleased

### CODE REVIEW

**Comment by jesmrec** [comment]
@hosy i did a first try over this and i was not able to reproduce the expected behaviour. These are the steps i did:

1. Remove security from device: no pin, no face id, no fingerprints...
2. Set `Branding.plist`  file:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>passcode.enforced-by-device</key>
	<true/>
</dict>
</plist>
```

3. Install the app in real device via Xcode
4. Open app

Passcode is not asked at the beginning and it's posible to attach an account at the time that app and device's security is off.

What's missing on my side @hosy?

iPhone XR 16.5.1
`94accb748`

**Comment by hosy** [comment]
@jesmrec updated the code, please can you have a look again?

**Comment by jesmrec** [comment]
Problem reported above is fixed

**Comment by jesmrec** [comment]
## QA

**Device**: iPhoneXR 16.5.1<br>
**Server**: oCIS 3.0


adding to `Branding.plist`

```
<key>passcode.enforced-by-device</key>
<true/>
```

#### Lock NOT enforced in the app

 
|    | Device with protection | Device without protection |
| :-- | :-------: | :---------: | 
| **App with protection** | Passcode set :white_check_mark:| Passcode set :white_check_mark:|
| **App without protection** | Passcode not set and not asked :white_check_mark:| Passcode asked to set :white_check_mark: |

#### Lock enforced in the app

Adding to `Branding.plist`

```
<key>passcode.enforced</key>
<true/>
```

|    | Device with protection | Device without protection |
| :-- | :-------: | :---------: | 
| **App with protection** |Passcode set :white_check_mark: |Passcode set :white_check_mark: |
| **App without protection** | Passcode set :white_check_mark: | Passcode set :white_check_mark:|


## Status changes

| Test Case | Steps | Expected Result | Result | Related Comment
| :-------- | :---- | :-------------- | :----: | :------------- |
| Set device protection I | 1. App not protected<br>2. Device not protected<br>3. Add protection to device<br>4. Open app| App does not ask to create new passcode.<br>Passcode is not removable from `Settings` | P m16|
| Set device protection II | 1. App protected<br>2. Device not protected<br>3. Add protection to device<br>4. Open app| App does not ask to create new passcode.<br>Passcode is removable from `Settings` | P m16|
| Unset device protection I |1. App not protected<br>2. Device protected<br>3. Remove  protection from device<br>4. Open app| App asks to create new passcode.<br>Passcode is not removable from `Settings` | P m16| 
| Unset device protection II |1. App protected<br>2. Device protected<br>3. Remove  protection from device<br>4. Open app| App does not ask to create new passcode.<br>Passcode is not removable from `Settings` | P m16|||
| Set app protection I | 1. App not protected<br>2. Device not protected<br>3 Open app| App asks to create new passcode.<br>Passcode is  not removable from `Settings` | P m16|
| Set app protection II | 1. App not protected<br>2. Device protected<br>3. Add protection to app<br>4. Open app| App does not ask to create new passcode.<br>Passcode is removable from `Settings` | P m16|
| Unset app protection I |1. App protected<br>2. Device protected<br>3. Remove protection from app<br>4. Open app| App does not ask to create new passcode. | P m16|
| Unset app protection II |1. App protected<br>2. Device not protected<br>3. Remove protection from app| Passcode is not removable from `Settings` | P m16|

Related: https://github.com/owncloud/ios-app/issues/532

File: ownCloudAppFramework/AppLock Settings/AppLockSettings.m [hunk 1]
```diff
@@ -53,6 +54,7 @@ + (OCClassSettingsIdentifier)classSettingsIdentifier
 {
 	return (@{
 		OCClassSettingsKeyPasscodeEnforced : @(NO),
+		OCClassSettingsKeyRequiredPasscodeDigits : @(NO),
```
Comment (felix-schwarz) [inline ownCloudAppFramework/AppLock Settings/AppLockSettings.m:57]: I think this should probably be `OCClassSettingsKeyPasscodeEnforcedByDevice` here.

File: ownCloudAppFramework/AppLock Settings/AppLockSettings.m [hunk 2]
```diff
@@ -80,7 +82,14 @@ + (OCClassSettingsMetadataCollection)classSettingsMetadata
 			OCClassSettingsMetadataKeyStatus	: OCClassSettingsKeyStatusAdvanced,
 			OCClassSettingsMetadataKeyCategory	: @"Passcode"
 		},
-
+		
+		OCClassSettingsKeyRequiredPasscodeDigits : @{
```
Comment (felix-schwarz) [inline ownCloudAppFramework/AppLock Settings/AppLockSettings.m:86]: Likewise, this should probably be for `OCClassSettingsKeyPasscodeEnforcedByDevice` here, too.

**Review by felix-schwarz — CHANGES_REQUESTED** [review]

**Comment by hosy** [comment]
@felix-schwarz fixed CR findings!

**Review by felix-schwarz — APPROVED** [review]

**Comment by jesmrec** [comment]
This is also QA-approved. Android & iOS clients totally alligned in their behaviours.

**Review by jesmrec — APPROVED** [review]

**Pull request merged**